lib/Runtime/Language/SourceTextModuleRecord.cpp

```diff
@@ -307,7 +307,7 @@ namespace Js
             OUTPUT_TRACE_DEBUGONLY(Js::ModulePhase, _u("\t>NotifyParentsAsNeeded\n"));
             NotifyParentsAsNeeded();
 
-            if (!WasDeclarationInitialized() && isRootModule)
+            if (!WasDeclarationInitialized() && (isRootModule || this->promise != nullptr))
```

Comment:
@rhuanjl thanks for fixing this code. I would like the test for this issue to be added so that we could catch future regression.  Could you add test for this issue? Preferably in test\es6\dynamic-module-functionality.js file.

Reply:
@akroshg I've written a test which I think covers this.

Concerned about the CI fails though, as the test it's failing is not entirely unrelated. This PR was to fix static imports that lead to dynamic imports - which were failing silently.

The CI test this is failing on is assessing dynamic imports which lead to static imports, and is apparently leading to ch segfaulting.

If I run the test individually with ch offline it passes - however if I run the whole framework offline it does fail, not sure what could of caused this - particularly confused as the individual test seems to run fine it's only in the context of the testing framework that it fails.

Reply:
you are hitting this assert 
script0
3>ASSERTION 2300: (D:\j\workspace\x86_debug_prtest06343b35\lib\Runtime\Language\SourceTextModuleRecord.cpp, line 806) moduleRecord->WasParsed()

could you run the test directly with ```-forcenative``` flag to see if this reproes?

Reply:
btw you could repro without forcenative as well, no?
```
core\Build\VcBuild\bin\x64_debug\ch.exe bug_issue_3257.js
script0
ASSERTION 24864: (C:\cg\core\lib\Runtime\Language\SourceTextModuleRecord.cpp, line 806) moduleRecord->WasParsed()
 Failure: (moduleRecord->WasParsed())
FATAL ERROR: ch.exe failed due to exception code c0000420
```

Reply:
Sorry I was testing with a test build switching to a debug build I now see this assert fail.

Having looked into this further I think my fix may have introduced a different error - I'll need to go through this again in more detail - oddly it runs fine for me with a TestBuild or a Release build but fails the assert with a Debug build and if I comment out the assert the Debug build Segfaults.

I'll endeavour work out I'm missing and fix later this week or early next; thanks for the help.

Reply:
Ok Two actions required before this can be merged, posting this now Because I would like some input on point 1:
1. Need to understand the correct fix to issue number 3306 - this change is reversing part of a previous fix to that issue - I think there is another fix needed that was hidden by the change I'm reversing (more below).

2. I need to learn how your test framework works my attempts so far have failed.

**More detail on point 1**
My fix is actually reversing part of this previous fix: https://github.com/Microsoft/ChakraCore/issues/3306

The fix posted for that issue involved removing the case I've added back - this suppressed the test failure but left no path for the host to be notified about a dynamic module that was not a root module.

The specific test failure I was seeing, bug_issue_3257, (also referenced in 3306) was due to CC attempting to execute a module that it didn't believe had been parsed.

**Process of the test - prior to any edit:**
1. Load a module which loads other modules staticly (these are all loaded and the host notified for them)
2. Load a script which dynamicly loads 2 of the already loaded modules

Execution process for point 2:
a) ch normalises the names of the modules noting it has already loaded modules by those names - returns the same specifier
b) the host is not asked to fetch them as there is already a specifier for a module that has been fetched
c) the pre-existing specifier is a child-specifier as they were previously loaded through static imports from a different module
d) the host is never notified that they're ready because the impact of issue 3902 is that the host is never notified for child dynamic imports

**Result of just adding in my bugfix for ChakraCore:**
Point(d) changes to:
The host is notified for the dynamic modules -> Assert failure (or segfault depending on build) as it can't find these modules

**Result of editing out the static imports in the test**
- The first module import is now the dynamic import
- ch throws an error "file not found" as it is attempting to evaluate them relative to the CWD - whereas they're given as relative to the script file

**Conclusion**
- currently Ch appears to correctly use paths relative to modules for static module imports 
- BUT ch uses paths relative to CWD for dynamic module imports from a script
- I suppressed this error by changing the script in the test to use paths relative to the CWD
-> The correct fix is probably to update ch's path handling